### PR TITLE
add segment voltage to can_message and analyzer

### PR DESCRIPTION
**Pull request by fregely**
## Changes

Added segment voltages that then can get sent to nero

## Notes

_Any other notes go here_

## Test Cases

- Case A
- Edge case
- ...

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please reach out to your Project Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes # (issue #)

### CODE REVIEW

File: Core/Src/analyzer.c [hunk 1]
```diff
@@ -340,6 +341,18 @@ void calc_pack_voltage_stats(acc_data_t *bmsdata)
 			total_volt += bmsdata->chip_data[c].cell_voltages[cell];
 			total_ocv +=
 				bmsdata->chip_data[c].open_cell_voltage[cell];
+
+			total_seg_volt +=
+				bmsdata->chip_data[c].cell_voltages[cell];
+
+			/* only for NERO */
```
Comment (jr1221) [inline Core/Src/analyzer.c]: can remove this comment.  just an artifact of old

File: Core/Src/analyzer.c [hunk 2]
```diff
@@ -340,6 +341,18 @@ void calc_pack_voltage_stats(acc_data_t *bmsdata)
...
+			if (c % 2 == 0) {
+				bmsdata->segment_average_volts[c / 2] =
+					total_seg_volt /
+					((float)(NUM_CELLS_ALPHA +
+						 NUM_CELLS_BETA));
+				total_seg_volt = 0;
```
Comment (jr1221) [inline Core/Src/analyzer.c:387]: I dont think this should be here.  this is in the iterator for every cell. I think it should be in the per-chip iterator.

File: Core/Src/can_messages.c [hunk 3]
```diff
@@ -285,6 +285,39 @@ void send_cell_voltage_message(acc_data_t *bmsdata)
...
 }
+void send_segment_volt_message(acc_data_t *bmsdata)
+{
+	struct __attribute__((__packed__)) {
+		int8_t segment1_average_volt;
+		int8_t segment2_average_volt;
+		int8_t segment3_average_volt;
+		int8_t segment4_average_volt;
+		int8_t segment5_average_volt;
+		int8_t segment6_average_volt;
```
Comment (jr1221) [inline Core/Src/can_messages.c]: so sorry I didnt metion there are 5 segments this year :cry:.  my bad.

File: Core/Src/can_messages.c [hunk 4]
```diff
@@ -285,6 +285,39 @@ void send_cell_voltage_message(acc_data_t *bmsdata)
...
+
+	} segment_volt_msg_data;
+
+	segment_volt_msg_data.segment1_average_volt =
+		bmsdata->segment_average_volts[0];
```
Comment (jr1221) [inline Core/Src/can_messages.c]: we need to scale these numbers or else the float to int conversion will give a whole number average.  Some quick math shows we can only scale by *10 (6 bits), as if we scaled by *100 it would take 9 bits.  But this doesnt even count sign bit so definitely *10.

File: Core/Src/can_messages.c [hunk 5]
```diff
@@ -285,6 +285,39 @@ void send_cell_voltage_message(acc_data_t *bmsdata)
 
 	queue_can_msg(msg);
 }
+void send_segment_volt_message(acc_data_t *bmsdata)
```
Comment (jr1221) [inline Core/Src/can_messages.c:288]: please find a spot to call this function.  probably how segment temp averages is called is fine.

File: Core/Src/can_messages.c [hunk 6]
```diff
@@ -285,6 +285,39 @@ void send_cell_voltage_message(acc_data_t *bmsdata)
...
+{
+	struct __attribute__((__packed__)) {
+		int8_t segment1_average_volt;
+		int8_t segment2_average_volt;
```
Comment (jr1221) [inline Core/Src/can_messages.c]: these dont really need to be signed right?  the average voltage of the cells _should_ never be negative?

**Review by jr1221 — CHANGES_REQUESTED** [review]
A few things. 

 Also: So we need to scale the values up here to preserve decimals.  I did some math using [CanCalc.xlsx](https://northeastern.sharepoint.com/:x:/r/sites/northeasternelectricracing/Shared%20Documents/Electrical/Project%20Files/24A/CANDUMP/CanCalc.xlsx?d=w1cdf906fde744b68928f6d52a1cd24a7&csf=1&web=1&e=7KLlmc&nav=MTJfRTU5X3s0NEIxRDE5Qi04Mjk4LTQ4NEQtOTNENC0xN0QyNTY0MDU3REN9)?  (see E59, lmk if access denied).  You could either: Use the `bitstream.h` library and acheive hundredths precision by not making each value byte-aligned, or you could simply change it to be *10 and keep using the packed struct feature.  I recommend the former as it is a fun challenge, but up to you.

File: Core/Src/can_messages.c [hunk 7]
```diff
@@ -285,6 +285,32 @@ void send_cell_voltage_message(acc_data_t *bmsdata)
...
+	bitstream_add(&segment_volt_msg_data, bmsdata->segment_average_volts[0],
+		      13);
+	bitstream_add(&segment_volt_msg_data, bmsdata->segment_average_volts[1],
+		      13);
+	bitstream_add(&segment_volt_msg_data, bmsdata->segment_average_volts[2],
+		      13);
+	bitstream_add(&segment_volt_msg_data, bmsdata->segment_average_volts[3],
+		      13);
+	bitstream_add(&segment_volt_msg_data, bmsdata->segment_average_volts[4],
+		      13);
```
Comment (jr1221) [inline Core/Src/can_messages.c]: 13*5 = 64, and the maximum message size is 64 bits.  we need to switch to 12.

File: Core/Inc/can_handler.h [hunk 8]
```diff
@@ -34,6 +34,8 @@
 #define CELL_TEMP_SIZE	    8
 #define SEGMENT_TEMP_CANID  0x85
 #define SEGMENT_TEMP_SIZE   6
+#define SEGMENT_VOLT_CANID  0x90
+#define SEGMENT_VOLT_SIZE   6
```
Comment (jr1221) [inline Core/Inc/can_handler.h]: this is the size in bytes. it cannot exceed 8.

**Review by jr1221 — CHANGES_REQUESTED** [review]
the message is a little too big.

**Review by jr1221 — DISMISSED** [review]
go ahead and merge.  good stuff.
